python/lsst/daf/butler/direct_query_driver/_sql_builders.py

```diff
@@ -0,0 +1,678 @@
+            sql_from_clause = (
+                self.select(postprocessing).cte() if cte else self.select(postprocessing).subquery()
+            )
+            return SqlJoinsBuilder(db=self.joins.db, from_clause=sql_from_clause).extract_columns(
+                self.columns, special=self.joins.special.keys()
+            )
+        return self.joins
+
+    def nested(
+        self, cte: bool = False, force: bool = False, *, postprocessing: Postprocessing | None
+    ) -> SqlSelectBuilder:
+        """Convert this builder into a `SqlSelectBuilder` that is guaranteed to
+        have no DISTINCT or GROUP BY, nesting it in a subquery or common table
+        expression only if needed to apply any current DISTINCT or GROUP BY
+        clauses.
+
+        This method consumes ``self``.
+
+        Parameters
+        ----------
+        cte : `bool`, optional
+            If `True`, nest via a common table expression instead of a
+            subquery.
+        force : `bool`, optional
+            If `True`, nest via a subquery or common table expression even if
+            there is no DISTINCT or GROUP BY.
+        postprocessing : `Postprocessing`
+            Struct representing post-query processing in Python, which may
+            require additional columns in the query results.
+
+        Returns
+        -------
+        builder : `SqlSelectBuilder`
+            `SqlSelectBuilder` with at least all columns in `columns`
+            available.  This may or may not be the `builder` attribute of this
+            object.
+        """
+        return SqlSelectBuilder(
+            self.into_from_builder(cte=cte, force=force, postprocessing=postprocessing), columns=self.columns
+        )
+
+    def union_subquery(
+        self, others: Iterable[SqlSelectBuilder], postprocessing: Postprocessing | None = None
+    ) -> SqlJoinsBuilder:
+        """Combine this builder with others to make a SELECT UNION subquery.
+
+        Parameters
+        ----------
+        others : `~collections.abc.Iterable` [ `SqlSelectBuilder` ]
+            Other query builders to union with.  Their `columns` attributes
+            must be the same as those of ``self``.
+        postprocessing : `Postprocessing`
+            Struct representing post-query processing in Python, which may
+            require additional columns in the query results.
+
+        Returns
+        -------
+        joins_builder : `SqlJoinsBuilder`
+            `SqlJoinsBuilder` with at least all columns in `columns` available.
+            This may or may not be the `joins` attribute of this object.
+        """
+        select0 = self.select(postprocessing)
+        other_selects = [other.select(postprocessing) for other in others]
+        return SqlJoinsBuilder(
+            db=self.joins.db,
+            from_clause=select0.union(*other_selects).subquery(),
+        ).extract_columns(self.columns, postprocessing)
+
+
+@dataclasses.dataclass(kw_only=True)
+class SqlColumns:
```

Comment:
The need to pull this out happened pretty late in the refactor and I just wasn't looking for further refinements at that stage.  I agree composition at least intuitively feels a little better, though I could also imagine that just adding a lot of `.columns` visual noise in practice, since the usage of the base class on its own is pretty niche.

Reply:
Punting this one to future tickets.